sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala

```diff
@@ -1001,6 +1001,9 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product with Tre
     case t: Seq[_] if t.forall(_.isInstanceOf[TreeNode[_]]) ||
       t.forall(_.isInstanceOf[Partitioning]) || t.forall(_.isInstanceOf[DataType]) =>
       JArray(t.map(parseToJson).toList)
+    // Scan Seq[Product] to make sure essential information can be written out.
+    case t: Seq[_] if t.forall(_.isInstanceOf[Product]) =>
```

Comment:
shall we merge it into the previous case? `case t: Seq[_] if t.forall(_.isInstanceOf[TreeNode[_]]) || ...|| t.forall(_.isInstanceOf[Product])`

Reply:
Sure, make sense. Will merge it.

Reply:
done.